### PR TITLE
[pytest][qos][config] Added pytests for "config qos reload" commands

### PR DESCRIPTION
**Pull request by smaheshm**
**- What I did**
Added pytests for "config qos reload" commands for single and multi ASIC platforms.

**- How I did it**
Since 'config qos reload' command invokdes'sonic-cfggen' process, add a "--dry-run" option to 'config qos reload' command so that it passes the right parameters to 'sonic-cfggen' to not 'read and write' to config DB in unit test env. Also added '--json-data' option that gives flexibility to test variable parameters like tiers and switch roles.

**- How to verify it**

Unit test.

```
======================================
samaddik@83f02bcc27aa:/sonic/src/sonic-utilities/tests$ pytest-3  -s -k reload
============================= test session starts ==============================
platform linux -- Python 3.7.3, pytest-3.10.1, py-1.7.0, pluggy-0.8.0
rootdir: /sonic/src/sonic-utilities/tests, inifile: pytest.ini
plugins: cov-2.6.0
collected 337 items / 335 deselected

config_test.py SETUP
Saving output in /tmp/qos_config_output.json
0
Running command: /usr/local/bin/sonic-cfggen  --additional-data '{"DEVICE_METADATA": {"localhost": {}}}' -t /sonic/src/sonic-utilities/tests/qos_config_input/buffers.json.j2,/tmp/qos_config_output.json -t /sonic/src/sonic-utilities/tests/qos_config_input/qos.json.j2,/tmp/qos_config_output.json -y /sonic/src/sonic-utilities/tests/qos_config_input/sonic_version.yml

.TEARDOWN
SETUP
Saving output in /tmp/qos_config_output.json<0,1,2..>
0
Running command: /usr/local/bin/sonic-cfggen -n asic0 --additional-data '{"DEVICE_METADATA": {"localhost": {}}}' -t /sonic/src/sonic-utilities/tests/qos_config_input/0/buffers.json.j2,/tmp/qos_config_output.json0 -t /sonic/src/sonic-utilities/tests/qos_config_input/0/qos.json.j2,/tmp/qos_config_output.json0 -y /sonic/src/sonic-utilities/tests/qos_config_input/sonic_version.yml
Running command: /usr/local/bin/sonic-cfggen -n asic1 --additional-data '{"DEVICE_METADATA": {"localhost": {}}}' -t /sonic/src/sonic-utilities/tests/qos_config_input/1/buffers.json.j2,/tmp/qos_config_output.json1 -t /sonic/src/sonic-utilities/tests/qos_config_input/1/qos.json.j2,/tmp/qos_config_output.json1 -y /sonic/src/sonic-utilities/tests/qos_config_input/sonic_version.yml

.TEARDOWN


=================== 2 passed, 335 deselected in 1.51 seconds ===================
```

**- Previous command output (if the output of a command-line utility has changed)**

**- New command output (if the output of a command-line utility has changed)**

### CODE REVIEW

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
👋

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
Gentle reminder @neethajohn @abdosi

File: tests/mock_tables/mock_multi_asic.py [hunk 1]
```diff
@@ -14,6 +14,8 @@ def mock_is_multi_asic():
 def mock_get_namespace_list(namespace=None):
     return ['asic0', 'asic1']
 
+
 multi_asic.get_num_asics = mock_get_num_asics
 multi_asic.is_multi_asic = mock_is_multi_asic
 multi_asic.get_namespace_list = mock_get_namespace_list
+multi_asic.get_namespaces_from_linux = mock_get_namespace_list
```
Comment (neethajohn) [inline tests/mock_tables/mock_multi_asic.py:21]: why is this needed? the value is the same as L20

File: tests/config_test.py [hunk 2]
```diff
@@ -61,8 +65,10 @@ class TestLoadMinigraph(object):
     def setup_class(cls):
         os.environ['UTILITIES_UNIT_TESTING'] = "1"
         print("SETUP")
+        import config.main
+        imp.reload(config.main)
```
Comment (neethajohn) [inline tests/config_test.py:69]: why do you need to reload? Not sure I understand the purposes of the reloads below as well. What is changing here so as to reload the config module?

File: tests/config_test.py [hunk 3]
```diff
@@ -95,3 +101,92 @@ def test_load_minigraph_with_disabled_telemetry(self, get_cmd_module, setup_sing
...
+        )
+        print(result.exit_code)
+        print(result.output)
+        assert result.exit_code == 0
+
+        cwd = os.path.dirname(os.path.realpath(__file__))
+        expected_result = os.path.join(
+            cwd, "qos_config_input", "config_qos.json"
+        )
+        assert filecmp.cmp(output_file, expected_result, shallow=False)
```
Comment (neethajohn) [inline tests/config_test.py:136]: Please validate the buffer template as well for single and masic . That also gets generated in the 'qos reload' command.

File: tests/mock_tables/mock_multi_asic.py (hunk 1 above)
Comment (smaheshm) [reply tests/mock_tables/mock_multi_asic.py:21]: Multi ASIC CLI uses this API so needs to be mocked.

File: tests/config_test.py (hunk 2 above)
Comment (smaheshm) [reply tests/config_test.py:69]: This is because of the multi ASIC variables that are set the first time module is loaded. Those variables should have different values based on single or multi ASIC. So reload ensures the test case has the right values.

File: tests/config_test.py (hunk 3 above)
Comment (smaheshm) [reply tests/config_test.py:136]: The main purpose of this test is to validate qos configs are generated for all ASICs on a mulit ASIC platform. 'sonic-cfgen' tests have the tests for qos and buffer templates.
https://github.com/Azure/sonic-buildimage/blob/master/src/sonic-config-engine/tests/test_j2files.py
" test_qos_dell6100_render_template, test_buffers_dell6100_render_template"

File: tests/conftest.py [hunk 4]
```diff
@@ -56,18 +56,44 @@ def get_cmd_module():
...
+def set_mock_apis():
+    import config.main as config
+    cwd = os.path.dirname(os.path.realpath(__file__))
+    device_info.get_paths_to_platform_and_hwsku_dirs = mock.MagicMock(
+        return_value=(
+            os.path.join(cwd, "."), os.path.join(cwd, "qos_config_input")
+        )
+    )
+    device_info.get_sonic_version_file = mock.MagicMock(
+        return_value=os.path.join(cwd, "qos_config_input/sonic_version.yml")
```
[the remark this answers is not in the document]
Comment (smaheshm) [reply tests/conftest.py:68]: makes sense, will separate them out.

File: tests/mock_tables/mock_multi_asic.py (hunk 1 above)
Comment (neethajohn) [reply tests/mock_tables/mock_multi_asic.py:21]: I was referring to this var 'get_namespaces_from_linux'. but saw its definition in multi_asic.py. So all good

Comment (smaheshm) [reply tests/mock_tables/mock_multi_asic.py:21]: me too :)

**Review by neethajohn — APPROVED** [review]

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by smaheshm** [comment]
retest this please

**Comment by lguohan** [comment]
this commit caused vkm test failure https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-image-pr/8207/

it is reverted, can you resubmit the pr?

**Comment by smaheshm** [comment]
> this commit caused vkm test failure https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-image-pr/8207/
> 
> it is reverted, can you resubmit the pr?

taking a look, will resubmit.

**Comment by smaheshm** [comment]
> > this commit caused vkm test failure https://sonic-jenkins.westus2.cloudapp.azure.com/job/vs/job/buildimage-vs-image-pr/8207/
> > it is reverted, can you resubmit the pr?
> 
> taking a look, will resubmit.

Added the offending API
https://github.com/Azure/sonic-buildimage/pull/6309
